### ISSUE TITLE
spec: clarify int type used for index values of range clauses

### ISSUE THREAD

**Issue by davecheney** [golang/go#18910]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?

go version devel +64c5280 Fri Feb 3 04:40:36 2017 +0000 linux/amd64

### What operating system and processor architecture are you using (`go env`)?

linux/amd64

### What did you do?
```
package main

import "fmt"

func main() {
        type int = int32
        var x int = 3
        for i := range []int{1, 2, 3} {
                fmt.Println(i == x)
        }
}
```
### What did you expect to see?
```
false
false
false
```
### What did you see instead?
```
# command-line-arguments
./alias.go:9: invalid operation: i == x (mismatched types int and int32)
```
https://golang.org/ref/spec#For_statements says that the type of the index variable is `int`, but it looks like this is the universe declaration of `int`, not the current alias.

**Comment by ianlancetaylor** [comment]
I think this is a spec bug.  I think it's correct for the for statement to use the universe declaration of `int`.

It's not an aliasing issue as such.  Your argument works just as well for `type int int32` as it does for `type int = int32`.

**Comment by davecheney** [comment]
> Your argument works just as well for type int int32 as it does for type int = int32.

Good point, well spotted.

**Comment by griesemer** [comment]
@ianlancetaylor Agreed. Changed title accordingly. That said, it's not clear to me what the best way to address this is in the spec. There's a lot of places where we talk about type 'int' and we always mean universe's 'int'.

**Comment by mdempsky** [comment]
Perhaps a paragraph under "Predeclared identifiers" mentioning that whenever the spec names a predeclared identifier, that it refers to the universal declaration?

**Comment by griesemer** [comment]
I've looked at this a bit. Adding a paragraph to the "Predeclared identifiers" section may help if one reads the spec sequentially and keeps that paragraph in mind at all times.

But that's not what people do when they have a specific question. Instead they look for the respective section in the spec. The clarification needs to be in those sections, if at all.

There's many mentions of the type `int`, and many other types as well (`rune`, `string`, etc.), all over the spec. We don't want to sprinkle the text with "predeclared `int` type" etc. Another option might be to link all these types to the "Predeclared identifiers" section. That will only help if people click on the link.

But more generally, we've now gone for almost 8 years and this has never come up as a problem: Readers tend to think of the predeclared types when we mention `int` type, `string` type, etc. It really only leads to confusion if people define their own names that happen to shadow predeclared identifiers and even that it needs a somewhat contrived example (as is the case in the example used for the issue).

We have a more wider issue related to confusing nomenclature. Closing this one in favor of  #4532.
